pkg/controller/podgc/gc_controller_test.go

```diff
@@ -135,90 +150,203 @@ func TestGCTerminated(t *testing.T) {
+			pods: []podMeta{
+				{name: "a", nodeName: "node", phase: v1.PodFailed},
+			},
+			deletedPodNames: sets.NewString(),
+		},
+		{
+			name:               "node deleted from client after quarantine",
+			initialClientNodes: []string{"node"},
+			delay:              2 * quarantineTime,
+			clientChanges:      []nodeEvent{{"node", watch.Deleted}},
+			pods: []podMeta{
+				{name: "a", nodeName: "node", phase: v1.PodFailed},
+			},
+			deletedPodNames: sets.NewString("a"),
+		},
+		{
+			name:               "node delete event after quarantine",
+			delay:              2 * quarantineTime,
+			initialListerNodes: []string{"node"},
+			nodeWatchEvents:    []nodeEvent{{"node", watch.Deleted}},
+			pods: []podMeta{
+				{name: "a", nodeName: "node", phase: v1.PodSucceeded},
+			},
+			deletedPodNames: sets.NewString(),
+		},
 	}
 
-	for i, test := range testCases {
-		client := fake.NewSimpleClientset()
-		gcc, podInformer := NewFromClient(client, test.threshold)
-		deletedPodNames := make([]string, 0)
-		var lock sync.Mutex
-		gcc.deletePod = func(_, name string) error {
-			lock.Lock()
-			defer lock.Unlock()
-			deletedPodNames = append(deletedPodNames, name)
-			return nil
-		}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			nodeList := &v1.NodeList{}
+			for _, nodeName := range test.initialClientNodes {
+				nodeList.Items = append(nodeList.Items, *testutil.NewNode(nodeName))
+			}
+			client := fake.NewSimpleClientset(nodeList)
+			gcc, podInformer, nodeInformer := NewFromClient(client, -1)
+			fakeClock := clock.NewFakeClock(time.Now())
+			gcc.nodeQueue = NewBasicUniqueWorkQueue(fakeClock)
 
-		creationTime := time.Unix(0, 0)
-		for _, pod := range test.pods {
-			creationTime = creationTime.Add(1 * time.Hour)
-			podInformer.Informer().GetStore().Add(&v1.Pod{
-				ObjectMeta: metav1.ObjectMeta{Name: pod.name, CreationTimestamp: metav1.Time{Time: creationTime}},
-				Status:     v1.PodStatus{Phase: pod.phase},
-				Spec:       v1.PodSpec{NodeName: "node"},
-			})
-		}
+			deletedPodNames := make([]string, 0)
+			var lock sync.Mutex
+			gcc.deletePod = func(_, name string) error {
+				lock.Lock()
+				defer lock.Unlock()
+				deletedPodNames = append(deletedPodNames, name)
+				return nil
+			}
 
-		pods, err := podInformer.Lister().List(labels.Everything())
-		if err != nil {
-			t.Errorf("Error while listing all Pods: %v", err)
-			return
-		}
-		gcc.gcOrphaned(pods)
+			for _, nodeName := range test.initialListerNodes {
+				nodeInformer.Informer().GetStore().Add(testutil.NewNode(nodeName))
+			}
+			creationTime := time.Unix(0, 0)
+			for _, pod := range test.pods {
+				creationTime = creationTime.Add(1 * time.Hour)
+				podInformer.Informer().GetStore().Add(&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{Name: pod.name, CreationTimestamp: metav1.Time{Time: creationTime}},
+					Status:     v1.PodStatus{Phase: pod.phase},
+					Spec:       v1.PodSpec{NodeName: pod.nodeName},
+				})
+			}
```

Comment:
I think the test would be simpler if you would test gcc.gcOrphaned (and the value would be the same).
What's the value of adding stuff to informers?

Reply:
gcOrhpaned depends on the up to date set of nodes and (possibly stale) state of lister.
I'm testing different combinations of both values before and after quarantine.
It could be done by passing lister state to gcOrhpaned directly, but that would still need 3-4 sets of nodes in tests.

Reply:
OK fair enough - that sounds reasonable to me.